Fix R004: ResourceData.Set() incompatible value type

Check: [R004](https://github.com/bflad/tfproviderlint/tree/master/passes/R004)

```
./gitlab/data_source_gitlab_user.go:198:22: R004: ResourceData.Set() incompatible value type: *time.Time
./gitlab/resource_gitlab_group_membership.go:154:22: R004: ResourceData.Set() incompatible value type: *github.com/xanzy/go-gitlab.ISOTime
./gitlab/resource_gitlab_service_slack.go:148:40: R004: ResourceData.Set() incompatible value type: func(b []byte) error
./gitlab/resource_gitlab_service_slack.go:149:38: R004: ResourceData.Set() incompatible value type: func(b []byte) error
```

We should fix these reported issues, and then remove this check from the list of ignored checks in the gitlabproviderlint make target.

To reproduce, run `go run ./gitlabproviderlint ./gitlab` (pending #432)

--- a/internal/provider/data_source_gitlab_project_protected_branch.go
+++ b/internal/provider/data_source_gitlab_project_protected_branch.go
@@ -98,12 +98,10 @@ func dataSourceGitlabProjectProtectedBranchRead(ctx context.Context, d *schema.R
 		return diag.Errorf("error getting protected branch (Project: %v / Name %v): %v", project, name, err)
 	}
 
-	// lintignore:R004 // TODO: Resolve this tfproviderlint issue
-	if err := d.Set("push_access_levels", convertBranchAccessDescriptionsToStateBranchAccessDescriptions(pb.PushAccessLevels)); err != nil {
+	if err := d.Set("push_access_levels", flattenBranchAccessDescriptions(pb.PushAccessLevels)); err != nil {
 		return diag.FromErr(err)
 	}
-	// lintignore:R004 // TODO: Resolve this tfproviderlint issue
-	if err := d.Set("merge_access_levels", convertBranchAccessDescriptionsToStateBranchAccessDescriptions(pb.MergeAccessLevels)); err != nil {
+	if err := d.Set("merge_access_levels", flattenBranchAccessDescriptions(pb.MergeAccessLevels)); err != nil {
 		return diag.FromErr(err)
 	}
 	if err := d.Set("allow_force_push", pb.AllowForcePush); err != nil {
@@ -118,33 +116,19 @@ func dataSourceGitlabProjectProtectedBranchRead(ctx context.Context, d *schema.R
 	return nil
 }
 
-type stateBranchAccessDescription struct {
-	AccessLevel            string `json:"access_level" mapstructure:"access_level"`
-	AccessLevelDescription string `json:"access_level_description" mapstructure:"access_level_description"`
-	GroupID                int    `json:"group_id,omitempty" mapstructure:"group_id,omitempty"`
-	UserID                 int    `json:"user_id,omitempty" mapstructure:"user_id,omitempty"`
-}
-
-func convertBranchAccessDescriptionsToStateBranchAccessDescriptions(descriptions []*gitlab.BranchAccessDescription) []stateBranchAccessDescription {
-	result := make([]stateBranchAccessDescription, 0)
-
+func flattenBranchAccessDescriptions(descriptions []*gitlab.BranchAccessDescription) (values []map[string]interface{}) {
 	for _, description := range descriptions {
-		result = append(result, convertBranchAccessDescriptionToStateBranchAccessDescription(description))
+		v := map[string]interface{}{
+			"access_level":             accessLevelValueToName[description.AccessLevel],
+			"access_level_description": description.AccessLevelDescription,
+		}
+		if description.UserID != 0 {
+			v["user_id"] = description.UserID
+		}
+		if description.GroupID != 0 {
+			v["group_id"] = description.GroupID
+		}
+		values = append(values, v)
 	}
-
-	return result
-}
-
-func convertBranchAccessDescriptionToStateBranchAccessDescription(description *gitlab.BranchAccessDescription) stateBranchAccessDescription {
-	stateDescription := stateBranchAccessDescription{
-		AccessLevel:            accessLevelValueToName[description.AccessLevel],
-		AccessLevelDescription: description.AccessLevelDescription,
-	}
-	if description.UserID != 0 {
-		stateDescription.UserID = description.UserID
-	}
-	if description.GroupID != 0 {
-		stateDescription.GroupID = description.GroupID
-	}
-	return stateDescription
+	return values
 }

--- a/internal/provider/data_source_gitlab_project_protected_branches.go
+++ b/internal/provider/data_source_gitlab_project_protected_branches.go
@@ -57,15 +57,6 @@ var _ = registerDataSource("gitlab_project_protected_branches", func() *schema.R
 	}
 })
 
-type stateProtectedBranch struct {
-	ID                        int                            `json:"id,omitempty" mapstructure:"id,omitempty"`
-	Name                      string                         `json:"name,omitempty" mapstructure:"name,omitempty"`
-	PushAccessLevels          []stateBranchAccessDescription `json:"push_access_levels,omitempty" mapstructure:"push_access_levels,omitempty"`
-	MergeAccessLevels         []stateBranchAccessDescription `json:"merge_access_levels,omitempty" mapstructure:"merge_access_levels,omitempty"`
-	AllowForcePush            bool                           `json:"allow_force_push,omitempty" mapstructure:"allow_force_push,omitempty"`
-	CodeOwnerApprovalRequired bool                           `json:"code_owner_approval_required,omitempty" mapstructure:"code_owner_approval_required,omitempty"`
-}
-
 func dataSourceGitlabProjectProtectedBranchesRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*gitlab.Client)
 
@@ -78,7 +69,7 @@ func dataSourceGitlabProjectProtectedBranchesRead(ctx context.Context, d *schema
 		return diag.FromErr(err)
 	}
 
-	allProtectedBranches := make([]stateProtectedBranch, 0)
+	var allProtectedBranches []*gitlab.ProtectedBranch
 	totalPages := -1
 	opts := &gitlab.ListProtectedBranchesOptions{}
 	for opts.Page = 0; opts.Page != totalPages; opts.Page++ {
@@ -88,20 +79,10 @@ func dataSourceGitlabProjectProtectedBranchesRead(ctx context.Context, d *schema
 			return diag.FromErr(err)
 		}
 		totalPages = resp.TotalPages
-		for _, pb := range pbs {
-			allProtectedBranches = append(allProtectedBranches, stateProtectedBranch{
-				ID:                        pb.ID,
-				Name:                      pb.Name,
-				PushAccessLevels:          convertBranchAccessDescriptionsToStateBranchAccessDescriptions(pb.PushAccessLevels),
-				MergeAccessLevels:         convertBranchAccessDescriptionsToStateBranchAccessDescriptions(pb.MergeAccessLevels),
-				AllowForcePush:            pb.AllowForcePush,
-				CodeOwnerApprovalRequired: pb.CodeOwnerApprovalRequired,
-			})
-		}
+		allProtectedBranches = append(allProtectedBranches, pbs...)
 	}
 
-	// lintignore:R004 // TODO: Resolve this tfproviderlint issue
-	if err := d.Set("protected_branches", allProtectedBranches); err != nil {
+	if err := d.Set("protected_branches", flattenProtectedBranches(allProtectedBranches)); err != nil {
 		return diag.FromErr(err)
 	}
 
@@ -113,4 +94,18 @@ func dataSourceGitlabProjectProtectedBranchesRead(ctx context.Context, d *schema
 	d.SetId(fmt.Sprintf("%d-%d", projectObject.ID, h))
 
 	return nil
+}
+
+func flattenProtectedBranches(protectedBranches []*gitlab.ProtectedBranch) (values []map[string]interface{}) {
+	for _, protectedBranch := range protectedBranches {
+		values = append(values, map[string]interface{}{
+			"id":                           protectedBranch.ID,
+			"name":                         protectedBranch.Name,
+			"push_access_levels":           flattenBranchAccessDescriptions(protectedBranch.PushAccessLevels),
+			"merge_access_levels":          flattenBranchAccessDescriptions(protectedBranch.MergeAccessLevels),
+			"allow_force_push":             protectedBranch.AllowForcePush,
+			"code_owner_approval_required": protectedBranch.CodeOwnerApprovalRequired,
+		})
+	}
+	return values
 }

--- a/internal/provider/resource_gitlab_branch_protection.go
+++ b/internal/provider/resource_gitlab_branch_protection.go
@@ -169,30 +169,30 @@ func resourceGitlabBranchProtectionRead(ctx context.Context, d *schema.ResourceD
 	d.Set("project", project)
 	d.Set("branch", pb.Name)
 
-	pushAccessLevels := convertAllowedAccessLevelsToBranchAccessDescriptions(pb.PushAccessLevels)
-	if len(pushAccessLevels) > 0 {
-		if err := d.Set("push_access_level", pushAccessLevels[0].AccessLevel); err != nil {
+	if pushAccessLevel, err := firstValidAccessLevel(pb.PushAccessLevels); err == nil {
+		if err := d.Set("push_access_level", accessLevelValueToName[*pushAccessLevel]); err != nil {
 			return diag.Errorf("error setting push_access_level: %v", err)
 		}
+	} else {
+		return diag.Errorf("unable to get push_access_level: %v", err)
 	}
 
-	mergeAccessLevels := convertAllowedAccessLevelsToBranchAccessDescriptions(pb.MergeAccessLevels)
-	if len(mergeAccessLevels) > 0 {
-		if err := d.Set("merge_access_level", mergeAccessLevels[0].AccessLevel); err != nil {
+	if mergeAccessLevels, err := firstValidAccessLevel(pb.MergeAccessLevels); err == nil {
+		if err := d.Set("merge_access_level", accessLevelValueToName[*mergeAccessLevels]); err != nil {
 			return diag.Errorf("error setting merge_access_level: %v", err)
 		}
+	} else {
+		return diag.Errorf("unable to get merge_access_level: %v", err)
 	}
 
 	if err := d.Set("allow_force_push", pb.AllowForcePush); err != nil {
 		return diag.Errorf("error setting allow_force_push: %v", err)
 	}
 
-	// lintignore: R004 // TODO: Resolve this tfproviderlint issue
-	if err := d.Set("allowed_to_push", convertAllowedToToBranchAccessDescriptions(pb.PushAccessLevels)); err != nil {
+	if err := d.Set("allowed_to_push", flattenNonZeroBranchAccessDescriptions(pb.PushAccessLevels)); err != nil {
 		return diag.Errorf("error setting allowed_to_push: %v", err)
 	}
-	// lintignore: R004 // TODO: Resolve this tfproviderlint issue
-	if err := d.Set("allowed_to_merge", convertAllowedToToBranchAccessDescriptions(pb.MergeAccessLevels)); err != nil {
+	if err := d.Set("allowed_to_merge", flattenNonZeroBranchAccessDescriptions(pb.MergeAccessLevels)); err != nil {
 		return diag.Errorf("error setting allowed_to_merge: %v", err)
 	}
 
@@ -284,36 +284,31 @@ func schemaAllowedTo() *schema.Schema {
 	}
 }
 
-func convertAllowedAccessLevelsToBranchAccessDescriptions(descriptions []*gitlab.BranchAccessDescription) []stateBranchAccessDescription {
-	result := make([]stateBranchAccessDescription, 0)
-
+func firstValidAccessLevel(descriptions []*gitlab.BranchAccessDescription) (*gitlab.AccessLevelValue, error) {
 	for _, description := range descriptions {
 		if description.UserID != 0 || description.GroupID != 0 {
 			continue
 		}
-		result = append(result, stateBranchAccessDescription{
-			AccessLevel:            accessLevelValueToName[description.AccessLevel],
-			AccessLevelDescription: description.AccessLevelDescription,
-		})
+		return &description.AccessLevel, nil
 	}
 
-	return result
+	return nil, fmt.Errorf("no valid access level found")
 }
 
-func convertAllowedToToBranchAccessDescriptions(descriptions []*gitlab.BranchAccessDescription) []stateBranchAccessDescription {
-	result := make([]stateBranchAccessDescription, 0)
-
+// flattenNonZeroBranchAccessDescriptions flattens the list of branch access descriptions for the tf state.
+// only descriptions with non-zero user id and group id are included in the tf state.
+func flattenNonZeroBranchAccessDescriptions(descriptions []*gitlab.BranchAccessDescription) (values []map[string]interface{}) {
 	for _, description := range descriptions {
 		if description.UserID == 0 && description.GroupID == 0 {
 			continue
 		}
-		result = append(result, stateBranchAccessDescription{
-			AccessLevel:            accessLevelValueToName[description.AccessLevel],
-			AccessLevelDescription: description.AccessLevelDescription,
-			UserID:                 description.UserID,
-			GroupID:                description.GroupID,
+		values = append(values, map[string]interface{}{
+			"access_level":             accessLevelValueToName[description.AccessLevel],
+			"access_level_description": description.AccessLevelDescription,
+			"user_id":                  description.UserID,
+			"group_id":                 description.GroupID,
 		})
 	}
 
-	return result
+	return values
 }

